workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/LayeredDialogSessions.kt

```diff
@@ -117,71 +130,60 @@ public class LayeredDialogSessions private constructor(
   public fun update(
     overlays: List<Overlay>,
     viewEnvironment: ViewEnvironment,
     updateBase: (environment: ViewEnvironment) -> Unit
   ) {
-    val modalIndex = overlays.indexOfFirst { it is ModalOverlay }
-    val showingModals = modalIndex > -1
+    // Set up a ViewEnvironment with the single DialogCollator instance that handles
+    // this entire view hierarchy. See that class for details.
+    val envWithDialogManager =
+      viewEnvironment.establishDialogCollator(idInCollator, sessions)
+    val collator = envWithDialogManager[DialogCollator]
 
+    val showingModals = overlays.any { it is ModalOverlay }
     allowEvents = !showingModals
 
     updateBase(
-      if (showingModals) viewEnvironment + (CoveredByModal to true) else viewEnvironment
+      if (showingModals) envWithDialogManager + (CoveredByModal to true) else envWithDialogManager
     )
 
-    val envPlusBounds = viewEnvironment + OverlayArea(bounds)
-
-    // On each update we build a new list of the running dialogs, both the
-    // existing ones and any new ones. We need this so that we can compare
-    // it with the previous list, and see what dialogs close.
-    val updatedSessions = mutableListOf<DialogSession>()
-    val oldSessionIterator = sessions.iterator()
-
-    for ((i, overlay) in overlays.withIndex()) {
-      val covered = i < modalIndex
-      val dialogEnv = if (covered) envPlusBounds + (CoveredByModal to true) else envPlusBounds
-
-      // We can't insert new dialogs in front of existing ones b/c there is no
-      // API to control z order of windows. But we can at least close lower dialogs
-      // that disappear from the list while preserving those that remain -- that is,
-      // updating [overlay1, overlay2] to [overlay2, overlay3] should preserve the
-      // dialog that was already created for overlay2. So for each entry in overlays,
-      // we advance to the first compatible old session, if any, and reuse it.
-      // The ones that we skip will be discarded and dismissed.
-
-      val oldSessionOrNull = oldSessionIterator.firstCompatible(overlay)
-
-      updatedSessions += oldSessionOrNull?.also { it.holder.show(overlay, dialogEnv) }
-        ?: overlay.toDialogFactory(dialogEnv)
-          .buildDialog(overlay, dialogEnv, context)
-          .let { holder ->
-            holder.onUpdateBounds?.let { updateBounds ->
-              holder.dialog.maintainBounds(holder.environment) { b -> updateBounds(b) }
+    val envPlusBounds = envWithDialogManager + OverlayArea(bounds)
+    val updates = mutableListOf<DialogSessionUpdate>()
+    overlays.forEach { overlay ->
+      updates += DialogSessionUpdate(overlay) { oldSessions, covered ->
+        val dialogEnv = if (covered) envPlusBounds + (CoveredByModal to true) else envPlusBounds
+        val oldSessionOrNull = oldSessions.find(overlay)
+
+        oldSessionOrNull
+          ?.also { session -> session.show(overlay, dialogEnv) }
+          ?: overlay.toDialogFactory(dialogEnv)
```

Comment:
```kotlin
        oldSessionOrNull
          // If there is already a matching session, it may have been hidden b/c out of order.
          ?.also { session -> session.show(overlay, dialogEnv) }
          // This is a new Overlay, create and show its Dialog.
```